Document fetchTree

wentasah opened the pull request:
# Motivation

WiP commit originally included in #8491.

# Context
<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

thufschmitt reviewed (commented):
Big :+1: on the idea, though I'll defer to the doc team for reviewing the content

fricklerhandwerk reviewed (approved):
Please reformat to one sentence per line, as this will make changes in the future much less noisy.

fricklerhandwerk commented on src/libexpr/primops/fetchTree.cc:
```diff
@@ -193,10 +193,50 @@ static void prim_fetchTree(EvalState & state, const PosIdx pos, Value * * args,
+    .doc = R"(
+      Fetch a source tree or a plain file using one of the supported
+      backends. *args* can be an attribute set representation of
+      [flake reference](../command-ref/new-cli/nix3-flake.html#flake-references)
```
```suggestion
      [flake reference](@docroot@/command-ref/new-cli/nix3-flake.md#flake-references)
```

fricklerhandwerk commented on src/libexpr/primops/fetchTree.cc:
````diff
@@ -193,10 +193,50 @@ static void prim_fetchTree(EvalState & state, const PosIdx pos, Value * * args,
+          ```nix
+          builtins.fetchTree "https://example.com/"
+          ```
+
+      > **Note**
+      >
+      > This function requires the [`flakes` experimental feature](@docroot@/contributing/experimental-features.md#xp-feature-flakes) to be enabled and the requested flake reference to be locked in `flake.lock`.
````
@Ericson2314 is the automatic experimental admonition already merged?

Ericson2314 replied:
It is now!

Ericson2314 replied:
Let's fix this later.

fricklerhandwerk reviewed (approved):
This looks good. Not sure if the experimental guard can be already automated. If not, ready to merge from my side.

wentasah commented:
I'll get to finish this tomorrow. Sorry for delays.

wentasah commented:
I made the following changes:
- Formatted as one sentence per line.
- Added a sentence about locked input (previous mention of `flake.lock` was incorrect). The question is whether it should be called *locked* (as in the [source code](https://github.com/wentasah/nix/blob/f11445952fcf0f5fd3f827997591d43a41f2074a/src/libfetchers/fetchers.hh#L67-L71)) or *immutable* as in [a recent @edolstra PR](https://github.com/NixOS/nix/pull/8477). I'd prefer *immutable*, in which case, we should also change [this error message](https://github.com/wentasah/nix/blob/f11445952fcf0f5fd3f827997591d43a41f2074a/src/libexpr/primops/fetchTree.cc#L186).
- Changed the first example to use "locked" input.
